relay-explainer.md

```diff
@@ -0,0 +1,91 @@
+* If the sink can require the transmission of I-frames to resynchronize decoding, those requests must be made available to the inserting entity
+* If the source is capable of adapting its transmission rate, the handler taking data from it must be able to make the source aware of its current capacity (which may or may not be the same as the downstream capacity).
+* If the source is capable of producing I-frames on demand, the handler taking data from it must be able to make those requests appropriately
+
+It is harmful to the model if there are couplings between the source of encoded data and the sink for encoded data that are not exposed to the handler.
+
+
+## Sample code
+
+Let's take a scenario where P2P relays are used to forward frames. Depending solely on a local peer for the stream is not very reliable in this setup. For consistent, glitch-free, low latency streaming, a redundant PeerConnection delivering the same stream is also setup(`recvPc1` and `recvPc2`). A peer can then choose to relay this stream to the next peer in the network (`relayPc`). We pass the readable stream of the incoming PeerConnections to a `relayFrames` function which reads the frames and creates new frames with updated metadata that conforms to the sender PeerConnection's requirements (`getUnifiedMetadata`). We then write one of these frames(`isDuplicate`) to the relay peer PerConnection. `getUnifiedMetadata` and `isDuplicate` are application-specific functions.
+
+```
+// code in main.js file
+const worker = new Worker('worker.js');
+
+// Let recvPc1, recvPc2 be the receiving PCs. 
+recvPc{1|2}.ontrack = evt => {
+  evt.receiver.transform = new RTCRtpScriptTransform(worker, { name: "receiverTransform" });
+};
+
+
+// Let relayPc be the PC used to relay frames to the next peer. 
+worker.onmessage = evt => {
+  relayPc.replaceTrack(evt.data);
+};
+```
+
+```
+// code in worker.js file
+async function relayFrames(reader, writer, encodedSource) {
+    if(!reader || !writer || !encodedSource){
+      return;
+    }
+    while (true) {
+        const {frame, done} = await reader.read();
+        if (done) return;
+
+        let newFrame = new RTCRtpEncodedVideoFrame(frame, getUnifiedMetadata(frame));
+        if(!isDuplicate(newFrame)) {
+            encodedSource.enqueue(newFrame);
+        }
+        // Put the original frame back in the receiver PC
+        writer.write(frame);
+    }
+}
+
+// Code to instantiate reader and writer from the RTPReceiver and RTPSender.
+onrtctransform = (event) => {
+  if (event.transformer.options.name == "receiverTransform") {
+    reader = event.transformer.readable;
+    writer = event.transformer.writable;
+    if (!encodedSource) {
+      encodedSource = new RTCRtpSenderEncodedSource();
+      postMessage(encodedSource.handle);
+    }
+  } else {
+    return;
+  }
+
+  relayFrames(reader, writer, encodedSource);
+};
+```
```

Comment:
But what about the racing? `onrtctransform` is going to fire twice here and free-run pumping frames from `recvPc1` and `recvPc2`, two independent peer connections, into the `transform` function in parallel. What syncs up these two independent sources? What assures these encoded frames have any relation? `isDuplicate`?

For fail-over, what's wrong with `replaceTrack`?

Reply:
Redo based on @youennf's [proposed API](https://github.com/w3c/webrtc-extensions/pull/198#issuecomment-2009454722):
```js
// main.js
const worker = new Worker('worker.js');
// Let recvPc1, recvPc2 be the receiving PCs. 
recvPc1.ontrack = ({receiver}) =>
  receiver.transform = new RTCRtpScriptTransform(worker, {name: "receiverTransform"});

recvPc2.ontrack = ({receiver}) =>
  receiver.transform = new RTCRtpScriptTransform(worker, {name: "receiverTransform"});

// Let relayPc be the PC used to relay frames to the next peer.
const [sender] = relayPc.getSenders();
const encodedSource = new RTCRtpSenderEncodedSource(worker);
await sender.replaceTrack(encodedSource);

// worker.js
let encodedWritable;
onrtcencodedsource = ({controller: {writable}}) => encodedWritable = writable;

onrtctransform = async ({transformer: {readable, writable, options}}) => {
  if (options.name != "receiverTransform") return;
  const [readable1, readable2] = readable.tee();
  await Promise.all([
    readable1.pipeTo(writable),
    readable2.pipeThrough(new TransformStream({transform})).pipeTo(encodedWritable)
  ]);

  function transform(frame, controller) {
    const newFrame = new RTCRtpEncodedVideoFrame(frame, getUnifiedMetadata(frame));
    if (!isDuplicate(newFrame)) {
      controller.enqueue(newFrame);
    }
  }
}
```
This doesn't solve my racing concerns, but I thought it might be helpful to show what the application code might look like with the different API shapes.

E.g. note this uses the `tee()` function which might run into https://github.com/whatwg/streams/issues/1156 since unlike WebCodecs our encoded chunks are not immutable (should be fine in this case since there's no transform on `readable1`, but it's brittle).